// @adminjs/sql/lib/dialects/base-database.parser.js
import KnexModule from 'knex';
const KnexConnection = KnexModule.knex;
export class BaseDatabaseParser {
    knex;
    dialect;
    connectionOptions;
    static dialects;
    constructor(dialect, connection) {
        if (!connection.database) {
            throw new Error('Please provide your database');
        }
        const knex = KnexConnection({
            client: dialect,
            connection,
        });
        this.dialect = dialect;
        this.connectionOptions = connection;
        this.knex = knex;
    }
    async parse() {
        throw new Error('Implement "parse" method for your database parser!');
    }
    async getSchema() {
        throw new Error('Implement "getSchema" method for your database parser!');
    }
    async getTables(schemaName) {
        throw new Error('Implement "getTables" method for your database parser!');
    }
    async getResources(tables, schemaName) {
        throw new Error('Implement "getResources" method for your database parser!');
    }
    async getProperties(table, schemaName) {
        throw new Error('Implement "getProperties" method for your database parser!');
    }
}
